theme: Enable style modifiers in theme.toml (#113)
helix-editor/helix: book/src/configuration.md

Styles in theme.toml are specified of in the form:

```toml
"name" = { fg = "#ffffff", bg = "#000000", modifiers = ["bold", "italic"] }
```
 where name represents what you want to style, fg specifies the foreground colour, bg the background colour, and modifiers is a list of style modifiers. bg or bold can be omitted to defer to the defaults.

 If you only want to specify the foreground color, you can do so as:

 ```toml
 "name" = "#ffffff"
 ```

 Possible entries for modifiers are:

| modifier |
------------
| bold |
| dim |
| italic |
| underlined |
| slow\_blink |
| rapid\_blink |
| reversed |
| hidden |
| crossed\_out |

 Possible entries for name are:

| name | notes |
----------------
| attribute | |
| keyword | |
| keyword.directive | preprocessor directives (\#if in C) |
| namespace | |
| punctuation | |
| punctuation.delimiter | |
| operator | |
| special | |
| property | |
| variable | |
| variable.parameter | |
| type | |
| type.builtin | |
| constructor | |

> | function | |

Reviewer: Another thing to note, this matches tree-sitter scopes. https://tree-sitter.github.io/tree-sitter/syntax-highlighting#theme

For a given highlight produced, styling will be determined based on the longest matching theme key.

We sorta half-follow the common scopes from here with some differences: https://macromates.com/manual/en/language_grammars

So it's enough to provide `function` to highlight `function.macro` and `function.builtin` as well, but you can use more specific scopes to highlight specific cases differently.

# Configuration

## Theme

Use a custom theme by placing a theme.toml in your config directory (i.e ~/.config/helix/theme.toml). The default theme.toml can be found [here](https://github.com/helix-editor/helix/blob/master/theme.toml), and user submitted themes [here](https://github.com/helix-editor/helix/blob/master/contrib/themes).

Styles in theme.toml are specified of in the form:

```toml

> "key" = { fg = "#ffffff", bg = "#000000", modifiers = ["bold", "italic"] }

Reviewer 2: ```suggestion
key = { fg = "#ffffff", bg = "#000000", modifiers = ["bold", "italic"] }
```
I think the `"` is a mistake is it?

Author: It's not a mistake, I followed the convention Blaz set in theme.toml. These are [quoted keys instead of bare keys](https://toml.io/en/v1.0.0#keys). I'm guessing we use quoted ones because the dotted ones won't parse correctly otherwise, but I'm not sure. @archseer can probably answer more definitively

Author: I don't agree about using bare keys. In my view if some keys need to be quoted, it's better to quote them all for consistency. Since you and @archseer are in agreement though, I'm deffering to your point of view. I've removed the quotes in these examples, and added a note about the need to quote keys with dots in them.

```

>  where name represents what you want to style, fg specifies the foreground colour, bg the background colour, and modifiers is a list of style modifiers. bg or bold can be omitted to defer to the defaults.
> 
>  If you only want to specify the foreground color, you can do so as:

Reviewer 2: ```suggestion

where `name` represents what you want to style, `fg` specifies the foreground color, `bg` the background color, and `modifiers` is a list of style modifiers. `bg` and `modifiers` can be omitted to defer to the defaults.

To specify only the foreground color:
```

 ```toml
 "key" = "#ffffff"
 ```

>  Possible entries for modifiers are:

Reviewer 2: ```suggestion
 Possible (term) modifiers:
```

| modifier |
------------
| bold |
| dim |
| italic |
| underlined |
| slow\_blink |
| rapid\_blink |
| reversed |
| hidden |
| crossed\_out |

>  Possible entries for key are:

Author: ```suggestion
 Possible keys:
```

| key | notes |
---------------
| attribute | |
| keyword | |
| keyword.directive | preprocessor directives (\#if in C) |
| namespace | |
| punctuation | |
| punctuation.delimiter | |
| operator | |
| special | |
| property | |
| variable | |
| variable.parameter | |
| type | |
| type.builtin | |
| constructor | |
| function | |
| function.macro | |
| function.builtin | |
| comment | |
| variable.builtin | |
| constant | |
| constant.builtin | |
| string | |
| number | |
| escape | escaped characters |
| label | used for lifetimes |
| module | |
| ui.background | |
| ui.linenr | |
| ui.statusline | |
| ui.popup | |
| ui.window | |
| ui.help | |
| ui.text | |
| ui.text.focus | |
| ui.menu.selected | |
| warning | LSP warning |
| error | LSP error |
| info | LSP info |
| hint | LSP hint |

> These keys match [tree-sitter scopes](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#theme). We sorta half-follow the common scopes from [here](https://macromates.com/manual/en/language_grammars) with some differences.

Reviewer 2: ```suggestion
These keys match [tree-sitter scopes](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#theme). We follow half of the the common scopes from [macromates language grammars](https://macromates.com/manual/en/language_grammars) with some differences.
```

Author: "Half-follow", has a different meaning to "follow half". I like the rest though, changing it to this:

```suggestion
These keys match [tree-sitter scopes](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#theme). We half-follow the common scopes from [macromates language grammars](https://macromates.com/manual/en/language_grammars) with some differences.
```

> For a given highlight produced, styling will be determined based on the longest matching theme key. So it's enough to provide function to highlight function.macro and function.builtin as well, but you can use more specific scopes to highlight specific cases differently.

Reviewer 2: ```suggestion
For a given highlight produced, styling will be determined based on the longest matching theme key. So it's enough to provide function to highlight `function.macro` and `function.builtin` as well, but you can use more specific scopes to highlight specific cases differently.
```

# Configuration

## Theme

Use a custom theme by placing a theme.toml in your config directory (i.e ~/.config/helix/theme.toml). The default theme.toml can be found [here](https://github.com/helix-editor/helix/blob/master/theme.toml), and user submitted themes [here](https://github.com/helix-editor/helix/blob/master/contrib/themes).

Styles in theme.toml are specified of in the form:

```toml
"key" = { fg = "#ffffff", bg = "#000000", modifiers = ["bold", "italic"] }
```

where `name` represents what you want to style, `fg` specifies the foreground color, `bg` the background color, and `modifiers` is a list of style modifiers. `bg` and `modifiers` can be omitted to defer to the defaults.

To specify only the foreground color:

 ```toml
 "key" = "#ffffff"
 ```

 Possible modifiers:

> | modifier |

Reviewer: Looking at the rich diff, these tables don't render for me correctly: https://github.com/helix-editor/helix/blob/f5883bbe7f3bf854c274706a75b66f24fb7a6736/book/src/configuration.md
